// @skyway-sdk/core/dist/media/stream/remote/media.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteMediaStreamBase = void 0;
const share_1 = require("../share");
const base_1 = require("./base");
class RemoteMediaStreamBase extends base_1.RemoteStreamBase {
    constructor(id, contentType, track) {
        super(id, contentType);
        this.id = id;
        this.contentType = contentType;
        this.track = track;
    }
    get isEnabled() {
        return this.track.enabled;
    }
    /**@internal */
    setIsEnabled(b) {
        this.track.enabled = b;
    }
    /**
     * @description [english] Attach the stream to the element.
     * @description [japanese] streamをelementに適用する.
     */
    attach(element) {
        this._element = element;
        (0, share_1.attachElement)(element, this.track);
    }
    /**
     * @description [english] Detach the stream from the element.
     * @description [japanese] elementからstreamを取り除く.
     */
    detach() {
        if (this._element) {
            (0, share_1.detachElement)(this._element, this.track);
            this._element = undefined;
        }
    }
}
exports.RemoteMediaStreamBase = RemoteMediaStreamBase;
